s3/pkg/service/multipart.go

```diff
@@ -444,8 +444,202 @@ func (s *s3Service) AbortMultipartUpload(ctx context.Context, in *pb.AbortMultip
+		return err
+	}
+	targetBucket, err := s.MetaStorage.GetBucket(ctx, targetBucketName, true)
+	if err != nil {
+		log.Errorln("get bucket failed with err:", err)
+		return err
+	}
+	isAdmin, tenantId, err := util.GetCredentialFromCtx(ctx)
+	if err != nil && isAdmin == false {
+		log.Error("get tenant id failed")
+		err = ErrInternalError
+		return err
+	}
+	if !isAdmin {
+		switch targetBucket.Acl.CannedAcl {
+		case "public-read-write":
+			break
+		default:
+			if targetBucket.TenantId != tenantId {
+				err = ErrBucketAccessForbidden
+				return err
+			}
+		} // TODO policy and fancy ACL
+	}
+
+	targetBackend, err := utils.GetBackend(ctx, s.backendClient, targetBucket.DefaultLocation)
+	if err != nil {
+		log.Errorln("failed to get backend client with err:", err)
+		return err
+	}
+	targetSd, err := driver.CreateStorageDriver(targetBackend.Type, targetBackend)
+	if err != nil {
+		log.Errorln("failed to create storage driver. err:", err)
+		return err
+	}
+	reader, err := srcSd.Get(ctx, srcObject.Object, offset, offset + size - 1)
+	if err != nil {
+		log.Errorln("failed to get data reader. err:", err)
+		return err
+	}
+	limitedDataReader := io.LimitReader(reader, size)
+
+	multipart, err := s.MetaStorage.GetMultipart(targetBucketName, targetObjectName, uploadId)
+	if err != nil {
+		log.Infoln("failed to get multipart. err:", err)
+		return err
+	}
+	result, err := targetSd.UploadPart(ctx, limitedDataReader, &pb.MultipartUpload{
+		Bucket:   targetBucketName,
+		Key:      targetObjectName,
+		UploadId: uploadId,
+		ObjectId: multipart.ObjectId},
+		partId, size)
+	if err != nil {
+		log.Errorln("failed to upload part to backend. err:", err)
+		return err
+	}
+	out.Etag = result.ETag
+	out.LastModified = time.Now().UTC().Unix()
+
+	log.Infoln("copy object part successfully.")
+	return nil
+}
+
 func (s *s3Service) ListObjectParts(ctx context.Context, in *pb.ListObjectPartsRequest, out *pb.ListObjectPartsResponse) error {
 	log.Info("ListObjectParts is called in s3 service.")
+	var err error
+	defer func() {
+		out.ErrorCode = GetErrCode(err)
+	}()
+
+	bucketName := in.BucketName
+	objectKey := in.ObjectKey
+	uploadId := in.UploadId
+
+	bucket, err := s.MetaStorage.GetBucket(ctx, bucketName, true)
+	if err != nil {
+		log.Errorln("get bucket failed with err:", err)
+		return err
+	}
+
+	multipart, err := s.MetaStorage.GetMultipart(bucketName, objectKey, uploadId)
+	if err != nil {
+		log.Errorln("failed to get multipart info. err:", err)
+		return err
+	}
+
+	isAdmin, tenantId, err := util.GetCredentialFromCtx(ctx)
+	if err != nil && isAdmin == false {
+		log.Error("get tenant id failed")
+		err = ErrInternalError
+		return err
+	}
+
+	if !isAdmin {
+		switch multipart.Metadata.Acl.CannedAcl {
+		case "public-read", "public-read-write":
+			break
+		case "authenticated-read":
+			if tenantId == "" {
+				err = ErrAccessDenied
+				return err
+			}
+		case "bucket-owner-read", "bucket-owner-full-controll":
```

Comment:
Does CannedAcl  include bucket-owner-read and bucket-owner-full-controll? Isn't owner always have full control?

Reply:
@sfzeng  yes, these two rules are member of CannedAcl.

Reply:
Ok, thanks.